### ISSUE TITLE
"Expected to fail" improvements

### ISSUE THREAD

**Issue by cgay** [dylan-lang/testworks#99]
Currently this feature is called `expected-failure?` and accepts a boolean or a function.

* I think it should allow a string value, to document why failure is expected. This could be a separate option, `expected-failure-reason`, for the case where a function is provided, but whatever we do I think a reason should be required.
* I would either call it `expect-failure?` or `expected-to-fail?`.

**Comment by cgay** [comment]
Further bike-shedding...

Unless we can find a really compelling reason to have it accept a function *and* a boolean *and* a string in the existing code base, let's replace the whole thing with `expected-failure-reason :: <string>`.

(Hint: there are zero uses of this feature in OD.)

**Comment by cgay** [comment]
This feature was added in b31b52e2fce85d1e5801d2d779af345aa0607d9c and mentions the obvious use case: to mark long-standing issues in the core libraries as expected to fail so their failures don't clutter the output and hide new failures, but can still be seen.

Note that there is no way to mark individual specs in `define interface-specification-suite` as expected to fail.

**Comment by cgay** [comment]
```
$ _build/bin/testworks-run --load libjson-test-suite.so --progress none
Loading library libjson-test-suite.so

testworks-run PASSED in 0.008022 seconds:
  Ran 1 suite: 1 passed (100.00000%)
  Ran 14 tests: 14 passed (100.00000%)
  Ran 0 benchmarks
  Ran 42 checks: 33 passed (78.571424%), 0 failed, 0 skipped, 0 not implemented, 9 crashed
```
That should say `Ran 14 tests: 13 passed, 1 expected failure`

(And frankly I don't think much of the other output is useful. This is what I'd prefer to see: `testworks-run PASSED in 0.008s: out of 14 tests, 13 passed, 1 expected failure`)

Also if I cause one of the tests to fail, the expected failures are then also printed in the summary, which is very inconsistent behavior.

```
_build/bin/testworks-run --load libjson-test-suite.so --progress none
Loading library libjson-test-suite.so

testworks-run failed
  test-print-false failed
    jprint(#f) = "alse" failed ["false" and "alse" are not =.  sizes differ (5 and 4), element 0 is the first non-matching element]
  test-parse-number-failures failed as expected
    123e3 => 123000 crashed [Error evaluating assertion expressions: Invalid floating point value.]
    123E3 => 123000 crashed [Error evaluating assertion expressions: Invalid floating point value.]
    123e+3 => 123000 crashed [Error evaluating assertion expressions: Invalid floating point value.]
    123E+3 => 123000 crashed [Error evaluating assertion expressions: Invalid floating point value.]
    123e-3 => 0.1230000d0 crashed [Error evaluating assertion expressions: Invalid floating point value.]
    123E-3 => 0.1230000d0 crashed [Error evaluating assertion expressions: Invalid floating point value.]
    123.123e3 => 123123 crashed [Error evaluating assertion expressions: Invalid floating point value.]
    123.1e-3 => 0.1231000d0 crashed [Error evaluating assertion expressions: Invalid floating point value.]
    -123.4e3 => -123400.00d0 crashed [Error evaluating assertion expressions: Invalid floating point value.]

testworks-run FAILED in 0.007139 seconds:
  Ran 1 suite: 0 passed (0.0%), 1 failed, 0 skipped, 0 not implemented, 0 crashed
  Ran 14 tests: 13 passed (92.857136%), 1 failed, 0 skipped, 0 not implemented, 0 crashed
  Ran 0 benchmarks
  Ran 42 checks: 32 passed (76.190472%), 1 failed, 0 skipped, 0 not implemented, 9 crashed
``` 

(Unrelated, but I think percentages in the display are pointless (no pun intended). They imply we're willing to have < 100% pass. We should make all tests pass or be marked expected to fail.)

**Comment by cgay** [comment]
The (or a) reason for allowing expected-to-fail to be a function is to provide a way to expect failure on one platform but not on others:

`define test test-x (expected-to-fail?: method () $os-name = #"win32" end)`

**Comment by cgay** [comment]
Remaining to do:

* Be consistent about showing the expected failures in the `failures-report-function`. I plan to remove them from the failures report since they can be seen in the progress output, the summary, and in the full report.

**Comment by cgay** [comment]
> Note that there is no way to mark individual specs in `define interface-specification-suite` as expected to fail.

Fixed in #129
